<?xml version="1.0"?>
<launch>
  <!-- Input topics to match exactly as published -->
  <arg name="input/img_front_left" default="/nuscenes/CAM_FRONT_LEFT/image"/>
  <arg name="input/img_front" default="/nuscenes/CAM_FRONT/image"/>
  <arg name="input/img_front_right" default="/nuscenes/CAM_FRONT_RIGHT/image"/>
  <arg name="input/img_back_left" default="/nuscenes/CAM_BACK_LEFT/image"/>
  <arg name="input/img_back" default="/nuscenes/CAM_BACK/image"/>
  <arg name="input/img_back_right" default="/nuscenes/CAM_BACK_RIGHT/image"/>

  <!-- CAN bus and scene topics -->
  <arg name="input/can_bus" default="/nuscenes/can_bus"/>
  <arg name="input/scene_token" default="/nuscenes/scene_tokens"/>

  <arg name="output/boxes" default="/tensorrt_bevformer/output/boxes"/>
  <arg name="output/bboxes" default="/nuscenes/bboxes"/>

  <arg name="input/img_front_left/camera_info" default="/nuscenes/CAM_FRONT_LEFT/camera_info"/>
  <arg name="input/img_front/camera_info" default="/nuscenes/CAM_FRONT/camera_info"/>
  <arg name="input/img_front_right/camera_info" default="/nuscenes/CAM_FRONT_RIGHT/camera_info"/>
  <arg name="input/img_back_left/camera_info" default="/nuscenes/CAM_BACK_LEFT/camera_info"/>
  <arg name="input/img_back/camera_info" default="/nuscenes/CAM_BACK/camera_info"/>
  <arg name="input/img_back_right/camera_info" default="/nuscenes/CAM_BACK_RIGHT/camera_info"/>

  <!-- Engine and ONNX paths -->
  <arg name="onnx_file" default="$(env HOME)/autoware_data/tensorrt_bevformer/bevformer_small.onnx"/>
  <arg name="engine_file" default="$(env HOME)/autoware_data/tensorrt_bevformer/bevformer_small.engine"/>
  <arg name="auto_convert" default="true"/>
  <arg name="workspace_size" default="4096"/>
  <arg name="precision" default="fp16"/>
  <arg name="debug_mode" default="false"/>

  <arg name="data_path" default="$(env HOME)/autoware_data"/>
  <arg name="model_name" default="bevformer_small"/>
  <arg name="model_config" default="$(find-pkg-share autoware_tensorrt_bevformer)/config/$(var model_name).yaml"/>
  <arg name="param_file" default="$(find-pkg-share autoware_tensorrt_bevformer)/config/bevformer.param.yaml"/>
  <arg name="plugin_path" default=""/>

  <!-- BEVFormer node -->
  <node pkg="autoware_tensorrt_bevformer" exec="autoware_tensorrt_bevformer_node" name="tensorrt_bevformer" output="screen">
    <!-- LOAD PARAMETER FILE FIRST -->
    <param from="$(var param_file)" allow_substs="true"/>

    <!-- THEN OVERRIDE SPECIFIC PARAMETERS -->
    <param name="model_params.engine_file" value="$(var engine_file)"/>
    <param name="model_params.onnx_file" value="$(var onnx_file)"/>
    <param name="model_params.auto_convert" value="$(var auto_convert)"/>
    <param name="model_params.workspace_size" value="$(var workspace_size)"/>
    <param name="model_params.plugin_path" value="$(var plugin_path)"/>
    <param name="model_params.precision" value="$(var precision)"/>
    <param name="post_process_params.debug_mode" value="$(var debug_mode)"/>

    <!-- ros topic name remapping -->
    <remap from="~/input/topic_img_fl" to="$(var input/img_front_left)"/>
    <remap from="~/input/topic_img_f" to="$(var input/img_front)"/>
    <remap from="~/input/topic_img_fr" to="$(var input/img_front_right)"/>
    <remap from="~/input/topic_img_bl" to="$(var input/img_back_left)"/>
    <remap from="~/input/topic_img_b" to="$(var input/img_back)"/>
    <remap from="~/input/topic_img_br" to="$(var input/img_back_right)"/>

    <remap from="~/input/can_bus" to="$(var input/can_bus)"/>
    <remap from="~/input/scene_token" to="$(var input/scene_token)"/>

    <remap from="~/input/topic_img_fl/camera_info" to="$(var input/img_front_left/camera_info)"/>
    <remap from="~/input/topic_img_f/camera_info" to="$(var input/img_front/camera_info)"/>
    <remap from="~/input/topic_img_fr/camera_info" to="$(var input/img_front_right/camera_info)"/>
    <remap from="~/input/topic_img_bl/camera_info" to="$(var input/img_back_left/camera_info)"/>
    <remap from="~/input/topic_img_b/camera_info" to="$(var input/img_back/camera_info)"/>
    <remap from="~/input/topic_img_br/camera_info" to="$(var input/img_back_right/camera_info)"/>

    <remap from="~/output/boxes" to="$(var output/boxes)"/>
    <remap from="~/output_bboxes" to="$(var output/bboxes)"/>
  </node>
</launch>
